extensions/amp-stream-gallery/0.1/amp-stream-gallery.js

```diff
@@ -0,0 +1,752 @@
+   * @private
+   */
+  getActionSource_(trust) {
+    return trust == ActionTrust.HIGH
+      ? ActionSource.GENERIC_HIGH_TRUST
+      : ActionSource.GENERIC_LOW_TRUST;
+  }
+
+  /**
+   * @return {number} The initial index for the carousel.
+   * @private
+   */
+  getInitialIndex_() {
+    const attr = this.element.getAttribute('slide') || '0';
+    return Number(getResponsiveAttributeValue(attr));
+  }
+
+  /**
+   * Determines how many whole items in addition to the current peek value can
+   * fit for a given item width. This can be rounded up or down to satisfy a
+   * max/min size constraint.
+   * @param {number} containerWidth The width of the container element.
+   * @param {number} itemWidth The width of each item.
+   * @return {number} The number of items to show.
+   */
+  getItemsForWidth_(containerWidth, itemWidth) {
+    const availableWidth = containerWidth - this.peek_ * itemWidth;
+    const fractionalItems = availableWidth / itemWidth;
+    const wholeItems = Math.floor(fractionalItems);
+    // Always show at least 1 whole item.
+    return Math.max(1, wholeItems) + this.peek_;
+  }
+
+  /**
+   * @return {number} The amount of horizontal space the arrows require. When
+   * the arrows are inset, this is zero as they do not take space.
+   * @private
+   */
+  getWidthTakenByArrows_() {
+    if (!this.outsetArrows_) {
+      return 0;
+    }
+
+    return (
+      this.prevArrowSlot_./* OK */ getBoundingClientRect().width +
+      this.nextArrowSlot_./* OK */ getBoundingClientRect().width
+    );
+  }
+
+  /**
+   * @param {!ActionSource|undefined} actionSource
+   * @return {boolean} Whether or not the action is a high trust action.
+   * @private
+   */
+  isHighTrustActionSource_(actionSource) {
+    return (
+      actionSource == ActionSource.WHEEL ||
+      actionSource == ActionSource.TOUCH ||
+      actionSource == ActionSource.GENERIC_HIGH_TRUST
+    );
+  }
+
+  /**
+   * TODO(sparhami) If swipe is disabled, then auto should show the inset arrow
+   * buttons, even if there is a peek value.
+   * @return {boolean}
+   * @private
+   */
+  shouldHideInsetButtons_() {
+    if (this.insetArrowVisibility_ == InsetArrowVisibility.ALWAYS) {
+      return false;
+    }
+
+    if (this.insetArrowVisibility_ == InsetArrowVisibility.NEVER) {
+      return true;
+    }
+
+    return this.hadTouch_ || this.peek_ != 0;
+  }
+
+  /**
+   *
+   * @param {number} peek
+   */
+  updatePeek_(peek) {
+    this.peek_ = Math.max(0, peek || 0);
+    this.updateVisibleCount_();
+  }
+
+  /**
+   *
+   * @param {number} maxItemWidth
+   */
+  updateMaxItemWidth_(maxItemWidth) {
+    this.maxItemWidth_ = maxItemWidth || Number.MAX_VALUE;
+    this.updateVisibleCount_();
+  }
+
+  /**
+   *
+   * @param {number} maxVisibleCount
+   */
+  updateMaxVisibleCount_(maxVisibleCount) {
+    this.maxVisibleCount_ = maxVisibleCount || Number.MAX_VALUE;
+    this.updateVisibleCount_();
+  }
+
+  /**
+   *
+   * @param {number} minItemWidth
```

Comment:
With these methods, is the `||` meant to catch a `0` being passed in or `undefined`/`null`?

Reply:
Both (kinda sorta). The type of the method's parameter is `{number}`, so the value is non-nullable (and cannot be undefined), but the step before turns `null` into 0 via a call to `Number(value)`.

Reply:
SGTM